consensus/istanbul/validator/default.go

```diff
@@ -141,133 +142,63 @@ func (valSet *defaultSet) List() []istanbul.Validator {
-
-	for i := uint64(0); i < valSet.subSize-2; i++ {
-		subset[i+2] = valSet.validators[indexs[i]]
-	}
-
-	if prevHash.Hex() == "0x0000000000000000000000000000000000000000000000000000000000000000" {
-		logger.Error("### subList", "prevHash", prevHash.Hex())
+	// TODO-Klaytn-Istanbul: investigate whether `valSet.GetProposer().Address()` is a proper value or the proposer should be calculated based on `view`
+	proposer := valSet.GetProposer()
+	if proposer == nil {
+		return valSet.List()
 	}
-
-	return subset
+	return valSet.SubListWithProposer(prevHash, proposer.Address(), view)
 }
 
-func (valSet *defaultSet) SubListWithProposer(prevHash common.Hash, proposer common.Address, view *istanbul.View) []istanbul.Validator {
+// SubListWithProposer composes a committee with given parameters.
+// The first member of the committee is set to the given proposer without calculating proposer with the given `view`.
+// The second member of the committee is calculated with a round number of the given view and `valSet.blockNum`.
+// The reset of the committee is selected with a random seed derived from `prevHash`.
+// This functions returns whole validators if it failed to compose a committee.
+func (valSet *defaultSet) SubListWithProposer(prevHash common.Hash, proposerAddr common.Address, view *istanbul.View) []istanbul.Validator {
 	valSet.validatorMu.RLock()
 	defer valSet.validatorMu.RUnlock()
 
-	if uint64(len(valSet.validators)) <= valSet.subSize {
-		return valSet.validators
-	}
-	hashstring := strings.TrimPrefix(prevHash.Hex(), "0x")
-	if len(hashstring) > 15 {
-		hashstring = hashstring[:15]
-	}
-	seed, err := strconv.ParseInt(hashstring, 16, 64)
-	if err != nil {
-		logger.Error("input", "hash", prevHash.Hex())
-		logger.Error("fail to make sub-list of validators", "seed", seed, "err", err)
-		return valSet.validators
-	}
-
-	// shuffle
-	subset := make([]istanbul.Validator, valSet.subSize)
-	subset[0] = New(proposer)
-	// next proposer
-	subset[1] = valSet.selector(valSet, subset[0].Address(), view.Round.Uint64())
+	validators := valSet.validators
+	validatorSize := uint64(len(validators))
+	committeeSize := valSet.subSize
 
-	proposerIdx, _ := valSet.GetByAddress(subset[0].Address())
-	nextproposerIdx, _ := valSet.GetByAddress(subset[1].Address())
+	// find the proposer
+	proposer := New(proposerAddr)
+	proposerIdx, _ := valSet.GetByAddress(proposer.Address())
 
-	// TODO-Klaytn-RemoveLater remove this check code if the implementation is stable.
-	if proposerIdx < 0 || nextproposerIdx < 0 {
-		vals := "["
-		for _, v := range valSet.validators {
-			vals += fmt.Sprintf("%s,", v.Address().Hex())
-		}
-		vals += "]"
-		logger.Error("idx should not be negative!", "proposerIdx", proposerIdx, "nextproposerIdx", nextproposerIdx, "proposer", subset[0].Address().Hex(),
-			"nextproposer", subset[1].Address().Hex(), "validators", vals)
+	// return early if the committee size is an edge value
+	if committeeSize == 1 {
+		return []istanbul.Validator{proposer}
```

Comment:
Why is an empty proposer with only address returned instead of a whole fetched proposer?

Reply:
Thank you! I changed the code to use the retuned object.